apollo-runtime/src/main/java/com/apollographql/apollo/internal/response/RealResponseWriter.kt

```diff
@@ -0,0 +1,225 @@
+      val fieldValue = value.value
+      if (fieldValue == null) {
+        fieldValues[fieldResponseName] = null
+      } else if (fieldValue is Map<*, *>) {
+        val nestedMap = rawFieldValues(fieldValue as Map<String, FieldDescriptor>)
+        fieldValues[fieldResponseName] = nestedMap
+      } else if (fieldValue is List<*>) {
+        fieldValues[fieldResponseName] = rawListFieldValues(fieldValue)
+      } else {
+        fieldValues[fieldResponseName] = fieldValue
+      }
+    }
+    return fieldValues
+  }
+
+  private fun rawListFieldValues(values: List<*>): List<*> {
+    val listValues = ArrayList<Any?>()
+    for (value in values) {
+      if (value is Map<*, *>) {
+        listValues.add(rawFieldValues(value as Map<String, FieldDescriptor>))
+      } else if (value is List<*>) {
+        listValues.add(rawListFieldValues(value))
+      } else {
+        listValues.add(value)
+      }
+    }
+    return listValues
+  }
+
+  private fun resolveFields(operationVariables: Operation.Variables,
+                            delegate: ResolveDelegate<Map<String, Any>?>, buffer: Map<String, FieldDescriptor>) {
+    val rawFieldValues = rawFieldValues(buffer)
+    for (fieldResponseName in buffer.keys) {
+      val fieldDescriptor = buffer[fieldResponseName]
+      val rawFieldValue = rawFieldValues[fieldResponseName]
+      delegate.willResolve(fieldDescriptor!!.field, operationVariables, fieldDescriptor.value)
+      when (fieldDescriptor.field.type) {
+        ResponseField.Type.OBJECT -> {
+          resolveObjectFields(fieldDescriptor, rawFieldValue as Map<String, Any>?, delegate)
+        }
+        ResponseField.Type.LIST -> {
+          resolveListField(fieldDescriptor.field, fieldDescriptor.value as List<*>?, rawFieldValue as List<*>?, delegate)
+        }
+        else -> {
+          if (rawFieldValue == null) {
+            delegate.didResolveNull()
+          } else {
+            delegate.didResolveScalar(rawFieldValue)
+          }
+        }
+      }
+      delegate.didResolve(fieldDescriptor.field, operationVariables)
+    }
+  }
+
+  private fun resolveObjectFields(fieldDescriptor: FieldDescriptor,
+                                  rawFieldValues: Map<String, Any>?, delegate: ResolveDelegate<Map<String, Any>?>) {
+    delegate.willResolveObject(fieldDescriptor.field, rawFieldValues)
+    val value = fieldDescriptor.value
+    if (value == null) {
+      delegate.didResolveNull()
+    } else {
+      resolveFields(operationVariables, delegate, value as Map<String, FieldDescriptor>)
+    }
+    delegate.didResolveObject(fieldDescriptor.field, rawFieldValues)
+  }
+
+  private fun resolveListField(listResponseField: ResponseField, fieldValues: List<*>?,
+                               rawFieldValues: List<*>?, delegate: ResolveDelegate<Map<String, Any>?>) {
+    if (fieldValues == null) {
+      delegate.didResolveNull()
+      return
+    }
+    for (i in fieldValues.indices) {
```

Comment:
This can be replaced with forEachWithIndex. Could be little bit better but this is also good.